### DISCUSSION TITLE
The previous error in onError

### DISCUSSION THREAD

**Discussion by andrewslock** [TanStack/query#1722] [Q&A]
I am not sure whether it's supposed to be so or it's a bag.

I am using a useQuery hook with my fetch function that can throw an error. I handle it like this:
`onError: (err) => {
if (err instanceof MyError) refetch();
    }`
But once the error has been thrown and I am trying to cancel the refetching query with `queryClient.cancelQueries()`  `onError` fires again with the same error so it breaks the logic of my handler.
Isn't it supposed to be CancelledError or something? Why does it return the same error gotten in the previous failed fetch?

**Comment by TkDodo** [comment]
If you want to refetch in case you get a certain error, you can implement the `retry` function.

**Reply by andrewslock** [reply]
Ok but the retry function cant be asynchronous (in case of some errors it must wait for the user to confirm that they want to refetch ). And still I dont understand why it gives me the same error

**Reply by TkDodo** [reply]
`refetch` is returned by `useQuery`, so if you use it in `onError` of the _same_ query - not sure how this would work 🙃 

have you tried `queryClient.refetchQueries` or so instead?

also, it would help if you'd provide a codesandbox example that shows the problem

**Reply by andrewslock** [reply]
Here's [the codesandbox example](https://codesandbox.io/s/the-same-error-8gcok)
First click the fetch query button and wait for it to fail. You will see that it throws "fetch error". Now, whenever you fetch the query and cancel it before it throws, onError will be instantly called with the same "fetch error".

**Reply by TkDodo** [reply]
I see. I thought this was maybe fixed with this PR: https://github.com/tannerlinsley/react-query/pull/1660

**Reply by andrewslock** [reply]
Well, on CodeSandbox I got 3.6.0 installed, so I guess it's not fixed 🤷
